File: C++/ScientificCalculator/src/Engine.cpp
#include "Engine.h"

#include "FunctionLibrary.h"
#include "Tokenizer.h"
#include "Parser.h"

#include <string>
#include <vector>

using namespace std;



// =======================
// Initialize
// =======================

void Engine::initialize()
{

    static bool initialized =
        false;


    if(initialized)
        return;


    FunctionLibrary::initialize();


    initialized = true;

}



// =======================
// Evaluate
// =======================

double Engine::evaluate(
    const string& expression
)
{

    initialize();


    Tokenizer tokenizer(
        expression
    );


    vector<string> tokens =
        tokenizer.tokenize();


    Parser parser(
        tokens
    );


    return parser.evaluate();

}

File: Java/ScientificCalculator/src/Engine.java
import java.util.List;

import parser.Tokenizer;
import parser.Parser;


public class Engine {


    // ==========================
    // Evaluate Expression
    // ==========================

    public static double evaluate(
            String expression
    ) {


        List<String> tokens =
                Tokenizer.tokenize(
                    expression
                );


        Parser parser =
                new Parser(
                    tokens
                );


        return parser.parse();

    }

}
